[AutoFill Debugging] Filter out hidden text when sanitizing text extraction results

whsieh opened the pull request:
#### a6cf92c2ad27fd663e7de0ec58dfd2288aecc322
<pre>
[AutoFill Debugging] Filter out hidden text when sanitizing text extraction results
<a href="https://bugs.webkit.org/show_bug.cgi?id=299736">https://bugs.webkit.org/show_bug.cgi?id=299736</a>
<a href="https://rdar.apple.com/161567763">rdar://161567763</a>

Reviewed by Aditya Keerthi and Abrar Rahman Protyasha.

Augment the text extraction filtering mechanism, such that it additionally snapshots long paragraphs
of text and verifies that the rendered text in the DOM is similar to what&apos;s visually legible.

Tests: TextExtractionTests.InteractionDebugDescription

* Source/WebCore/PAL/pal/cocoa/VisionSoftLink.mm:
* Source/WebCore/page/text-extraction/TextExtraction.cpp:
(WebCore::TextExtraction::rangeForExtractedText):

Add a helper method to map (visible text, optional node ID) to a `SimpleRange`. If unspecified, this
just searches the entire body for the text.

* Source/WebCore/page/text-extraction/TextExtraction.h:
* Source/WebCore/page/text-extraction/TextExtractionTypes.h:
* Source/WebKit/Platform/cocoa/ImageAnalysisUtilities.h:
* Source/WebKit/Platform/cocoa/ImageAnalysisUtilities.mm:
(WebKit::textRecognitionQueueSingleton):
(WebKit::recognizeText):

Add a helper method that takes a `CGImageRef`, and uses the Vision framework to scan the image for
all visible text. Most of the heavy lifting is done on a background queue, and the results are
dispatched back to the main runloop, where we invoke the completion handler.

* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _requestTextExtraction:completionHandler:]):
(-[WKWebView _validateText:inNode:completionHandler:]):

Implement the main internal helper method here. The results are cached until main frame navigation,
such that repeated requests to validate the same strings don&apos;t incur any snapshotting or OCR cost.
The optional node identifier here is optional, and we fall back to searching the entire document if
it&apos;s `nil`. Otherwise, it can make searching for the visible text a bit more efficient, and
accurate.

(-[WKWebView _clearTextExtractionFilterCache]):

Pull out common logic to reset caches related to text extraction filtering after process swap or
mainframe navigation.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _processWillSwapOrDidExit]):

`-_clearTextExtractionFilterCache` when process swapping.

* Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionUtilities.h:
* Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionUtilities.mm:
(WebKit::filterTextRecursive):
(WebKit::filterText):

Extend `filterText` to additionally OCR any long piece of text that has passed first-level
validation in `TextExtractionFilter`; for any piece of text that&apos;s too dissimilar (with an arbitrary
threshold of 0.5 edit distance) from its OCR results, we replace that text with its OCR text. This
ensures that in cases where OCR is a tiny bit off (mistaking a `0` for an `O`, or `1` or an `l`) we
just allow the original string through, but in cases where most of the text is missing or different,
we instead prefer OCR results. This leads to no change in behavior for most non-hidden text.

(WebKit::computeSimilarity):

Add a helper method to compute edit distance similarity between the two given strings, or
`std::nullopt` if the strings are too short (below the `minimumLength`) for a sensible comparison.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::takeSnapshotOfExtractedText):

Add an IPC message to snapshot an extracted text range, given text and node ID. This currently only
applies to main frame content, since text extraction only works with the main frame right now; when
we extend this extraction to subframes, we&apos;ll want to instead move this to `WebFrame(Proxy)` and
have `-_validateText:` above take a frame ID.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::processDidExit):
(WebKit::PageClientImpl::processWillSwap):
(WebKit::PageClientImpl::didCommitLoadForMainFrame):
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::processWillSwap):
(WebKit::PageClientImpl::didCommitLoadForMainFrame):

Make these all call into `-_clearTextExtractionFilterCache`.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::takeSnapshotOfExtractedText):

Use `rangeForExtractedText` above to map the text to a DOM range, and then use `TextIndicator` to
produce a snapshot of the range and send it back to the UI process.

* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/TextExtractionTests.mm:
(TestWebKitAPI::TEST(TextExtractionTests, InteractionDebugDescription)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/debug-text-extraction.html:

Augment an existing API test, to confirm that long runs of hidden text don&apos;t show up in the debug
text extraction output (in this case, white text on a white background).

Canonical link: <a href="https://commits.webkit.org/300716@main">https://commits.webkit.org/300716@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1236923167bcc67dd16900dd1efb49ec14ef0f4d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123607 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43322 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34018 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130374 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/75751 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/f4c7bd1c-9709-4d6e-b869-5f1442788baa) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/125484 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44045 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51916 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/93994 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/75751 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/8a84a0f6-eaf5-48a0-99be-0d99b28d3295) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126560 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35088 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110581 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74590 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/3f6955c5-aabc-4669-b877-630b313ef587) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34054 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28743 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/73854 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104808 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28963 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133068 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50558 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38493 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102460 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/50933 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106802 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102302 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26004 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47655 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25893 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/47369 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50412 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56174 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/49886 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53233 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51561 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ece5f48fcc7449e0cd38d2de1564ad82b6fc87a4. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51505

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/8d346a2bc9cdae0f1328f5093209456ab4928502. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51505

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6e9d4e02475e7a8e5ba99f14f8aad04465c484fe. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51505

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionUtilities.mm:
```diff
@@ -239,30 +240,102 @@ static WKTextExtractionEventListenerTypes eventListenerTypes(OptionSet<TextExtra
 
 #endif // ENABLE(TEXT_EXTRACTION_FILTER)
 
+std::optional<double> computeSimilarity(NSString *stringA, NSString *stringB, unsigned minimumLength)
```
I wonder if this should use `String` and be moved down into WTF at some point? Not necessary for now.

pxlcoder commented on Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
```diff
@@ -220,13 +220,25 @@ - (BOOL)_isHostedInAnotherProcess;
+    [webView _processDidExit];
+
+#if ENABLE(TEXT_EXTRACTION_FILTER)
+    [webView _clearTextExtractionFilterCache];
```
Maybe we can just call this in `_processWillSwapOrDidExit`?

pxlcoder commented on Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
```diff
@@ -220,13 +220,25 @@ - (BOOL)_isHostedInAnotherProcess;
+    [webView _processWillSwap];
+
+#if ENABLE(TEXT_EXTRACTION_FILTER)
+    [webView _clearTextExtractionFilterCache];
```
(see previous comment)

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/53ba75e90da21270eb6622ac29bbf2416d309e76. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51505

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/22492d990d9f80ed65ea0a51ac60b5c7a6755499. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51505

aprotyas commented on Source/WebKit/Platform/cocoa/ImageAnalysisUtilities.mm:
```diff
@@ -438,6 +438,69 @@ void requestPayloadForQRCode(CGImageRef image, CompletionHandler<void(NSString *
 
+static WorkQueue& textRecognitionQueueSingleton()
+{
+    static NeverDestroyed<Ref<WorkQueue>> queue {
```
```suggestion
    static NeverDestroyed queue {
```

aprotyas commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -6774,6 +6774,85 @@ - (void)_describeInteraction:(_WKTextExtractionInteraction *)wkInteraction compl
+            static constexpr auto minimumLength = 10;
+
+            auto similarity = WebKit::computeSimilarity(text.get(), recognizedText, minimumLength);
+            if (similarity && (*similarity < minimumSimilarity)) {
```
`std::optional<T>::operator<` does this for you btw

```suggestion
            if (similarity < minimumSimilarity) {
```

whsieh committed:
[AutoFill Debugging] Filter out hidden text when sanitizing text extraction results
https://bugs.webkit.org/show_bug.cgi?id=299736
rdar://161567763

Reviewed by Aditya Keerthi and Abrar Rahman Protyasha.

Augment the text extraction filtering mechanism, such that it additionally snapshots long paragraphs
of text and verifies that the rendered text in the DOM is similar to what's visually legible.

Tests: TextExtractionTests.InteractionDebugDescription

* Source/WebCore/PAL/pal/cocoa/VisionSoftLink.mm:
* Source/WebCore/page/text-extraction/TextExtraction.cpp:
(WebCore::TextExtraction::rangeForExtractedText):

Add a helper method to map (visible text, optional node ID) to a `SimpleRange`. If unspecified, this
just searches the entire body for the text.

* Source/WebCore/page/text-extraction/TextExtraction.h:
* Source/WebCore/page/text-extraction/TextExtractionTypes.h:
* Source/WebKit/Platform/cocoa/ImageAnalysisUtilities.h:
* Source/WebKit/Platform/cocoa/ImageAnalysisUtilities.mm:
(WebKit::textRecognitionQueueSingleton):
(WebKit::recognizeText):

Add a helper method that takes a `CGImageRef`, and uses the Vision framework to scan the image for
all visible text. Most of the heavy lifting is done on a background queue, and the results are
dispatched back to the main runloop, where we invoke the completion handler.

* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _requestTextExtraction:completionHandler:]):
(-[WKWebView _validateText:inNode:completionHandler:]):

Implement the main internal helper method here. The results are cached until main frame navigation,
such that repeated requests to validate the same strings don't incur any snapshotting or OCR cost.
The optional node identifier here is optional, and we fall back to searching the entire document if
it's `nil`. Otherwise, it can make searching for the visible text a bit more efficient, and
accurate.

(-[WKWebView _clearTextExtractionFilterCache]):

Pull out common logic to reset caches related to text extraction filtering after process swap or
mainframe navigation.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _processWillSwapOrDidExit]):

`-_clearTextExtractionFilterCache` when process swapping.

* Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionUtilities.h:
* Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionUtilities.mm:
(WebKit::filterTextRecursive):
(WebKit::filterText):

Extend `filterText` to additionally OCR any long piece of text that has passed first-level
validation in `TextExtractionFilter`; for any piece of text that's too dissimilar (with an arbitrary
threshold of 0.5 edit distance) from its OCR results, we replace that text with its OCR text. This
ensures that in cases where OCR is a tiny bit off (mistaking a `0` for an `O`, or `1` or an `l`) we
just allow the original string through, but in cases where most of the text is missing or different,
we instead prefer OCR results. This leads to no change in behavior for most non-hidden text.

(WebKit::computeSimilarity):

Add a helper method to compute edit distance similarity between the two given strings, or
`std::nullopt` if the strings are too short (below the `minimumLength`) for a sensible comparison.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::takeSnapshotOfExtractedText):

Add an IPC message to snapshot an extracted text range, given text and node ID. This currently only
applies to main frame content, since text extraction only works with the main frame right now; when
we extend this extraction to subframes, we'll want to instead move this to `WebFrame(Proxy)` and
have `-_validateText:` above take a frame ID.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::processDidExit):
(WebKit::PageClientImpl::processWillSwap):
(WebKit::PageClientImpl::didCommitLoadForMainFrame):
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::processWillSwap):
(WebKit::PageClientImpl::didCommitLoadForMainFrame):

Make these all call into `-_clearTextExtractionFilterCache`.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::takeSnapshotOfExtractedText):

Use `rangeForExtractedText` above to map the text to a DOM range, and then use `TextIndicator` to
produce a snapshot of the range and send it back to the UI process.

* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/TextExtractionTests.mm:
(TestWebKitAPI::TEST(TextExtractionTests, InteractionDebugDescription)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/debug-text-extraction.html:

Augment an existing API test, to confirm that long runs of hidden text don't show up in the debug
text extraction output (in this case, white text on a white background).

Canonical link: https://commits.webkit.org/300716@main
